_posts/2019-12-10-variational-gaussian-process/vgp.Rmd

```diff
@@ -0,0 +1,549 @@
+      super()$`__init__`(kwargs)
+      dtype <- kwargs[["dtype"]]
+      self$`_amplitude` = self$add_variable(initializer = initializer_zeros(),
+                                            dtype = dtype,
+                                            name = 'amplitude')
+      self$`_length_scale` = self$add_variable(initializer = initializer_zeros(),
+                                               dtype = dtype,
+                                               name = 'length_scale')
+      NULL
+    },
+    
+    call = function(self, x, ...) {
+      x
+    },
+    
+    kernel = bt$property(
+      reticulate::py_func(
+        function(self)
+          tfp$math$psd_kernels$ExponentiatedQuadratic(
+            amplitude = tf$nn$softplus(array(0.1) * self$`_amplitude`),
+            length_scale = tf$nn$softplus(array(2) * self$`_length_scale`)
+          )
+      )
+    )
+  )
+)
+```
+
+
+The Gaussian Process kernel used is one of several available in `tfp.math.psd_kernels` (`psd` standing for positive semidefinite), and probably the one that comes to mind first when thinking of GPR: the _squared exponential_, or _exponentiated quadratic_ ^[See David Duvenaud's https://www.cs.toronto.edu/~duvenaud/cookbook/ for an excellent synopsis of kernels and kernel composition.]. The version used in TFP, with hyperparameters _amplitude_ $a$ and _length scale_ $\lambda$, is
+
+$$k(x,x') = 2 \ a \ exp (\frac{- 0.5 (x−x')^2}{\lambda^2}) $$
+
+Here the interesting parameter is the length scale $\lambda$. When we have several features, their length scales -- as induced by the learning algorithm -- reflect their importance: If, for some feature, $\lambda$ is large, the respective squared deviations from the mean don't matter that much. The inverse length scale can thus be used for _automatic relevance determination_ [@Neal].
+
+The second thing to take care of is choosing the initial index points. From experiments, the exact choices don't matter that much, as long as the data are sensibly covered. For instance, an alternative way we tried was to construct an empirical distribution ([tfd_empirical](https://rstudio.github.io/tfprobability/reference/tfd_empirical.html)) from the data, and then sample from it. Here instead, we just use an -- unnecessary, admittedly, given the availability of `sample` in R -- fancy way to select random observations from the training data:
+
+```{r}
+num_inducing_points <- 50
+
+sample_dist <- tfd_uniform(low = 1, high = nrow(train) + 1)
+sample_ids <- sample_dist %>%
+  tfd_sample(num_inducing_points) %>%
+  tf$cast(tf$int32) %>%
+  as.numeric()
+sampled_points <- train[sample_ids, 1:8]
+```
+
+
+One interesting point to note before we start training. Computation of the posterior predictive parameters involves a Cholesky decomposition, which could fail if, due to numerical issues, the covariance matrix is no longer positive definite. A sufficient action to take in our case is to do all computations using `tf$float64`:
```

Comment:
```suggestion
One interesting point to note before we start training: Computation of the posterior predictive parameters involves a Cholesky decomposition, which could fail if, due to numerical issues, the covariance matrix is no longer positive definite. A sufficient action to take in our case is to do all computations using `tf$float64`:
```

Reply:
Switched it to a colon, since it reads like a fragment to me. Alternatively, you could expand it to be ~ _There is one more interesting..._

Reply:
ok! Maybe again a, I forget the acronym, English as second language thing...